Feature/speed limit pjtc extend controller

hslusarek opened the pull request:
PR to "correctly" integrate Cartesian velocity check, into `PilzJointTrajectoryController` (and its base class `JointTrajectoryController`).

martiniil commented on pilz_control/include/pilz_control/pilz_joint_trajectory_controller_impl.h:
```diff
@@ -178,6 +180,39 @@ triggerMovementToHoldPosition()
+reactToFailedStateCheck(const ros::Time& updated_uptime, const Trajectory& curr_traj)
+{
+  switchToHoldMode();
+  JointTrajectoryController::preemptActiveGoal();
```
I would not rely on actionlib to meet the "realtime" requirements of the update cycle.
In other words this should be executed by another thread. Maybe also the other lines included in `reactToFailedStateCheck`. At least we have to take a close look at it.

hslusarek replied:
Done.

martiniil commented on pilz_control/include/pilz_control/pilz_joint_trajectory_controller_impl.h:
```diff
@@ -178,6 +180,39 @@ triggerMovementToHoldPosition()
+
+template <class SegmentImpl, class HardwareInterface>
+inline void PilzJointTrajectoryController<SegmentImpl, HardwareInterface>::
+switchToHoldMode()
```
Why do you omit the lock used here:
> handleHoldRequest(std_srvs::TriggerRequest&, std_srvs::TriggerResponse& response)
{
  std::lock_guard\<std::mutex> lock(sync_mutex_);

 > if(active_mode_ == Mode::HOLD)

hslusarek replied:
First off: Thanks for pointing this out.
You actually made me aware of the fact that we are facing a much more general problem:
How should we deal with mutexes in the `update()` function?
Let's for example have look at [`sync_mutex_`](https://github.com/PilzDE/pilz_robots/blob/melodic-devel/pilz_control/include/pilz_control/pilz_joint_trajectory_controller.h#L129) in the `PilzJointTrajectoryController` class. The `sync_mutex_` mutex can be locked as long as it takes to perform a complete stop (see [`handleHoldRequest` function](https://github.com/PilzDE/pilz_robots/blob/melodic-devel/pilz_control/include/pilz_control/pilz_joint_trajectory_controller_impl.h#L111)). This can be a multiple of the actual controller period which, therefore, violates the "realtime" aspect if the mutex is used to change the value of the `active_mode_` member.
But even if I for the moment neglect the `sync_mutex_` mutex, let's have a look at the member [`curr_trajectory_box_`](https://github.com/ros-controls/ros_controllers/blob/melodic-devel/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller.h#L196) of the `JointTrajectoryController` base class. The `curr_trajectory_box_` member is e.g. needed to [set the stop trajectory](https://github.com/ros-controls/ros_controllers/blob/melodic-devel/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h#L811) in case of a failed state check. The only problem is, the member internally also uses a mutex. Who does guarantee that the member is not locked for to long. Is this guaranteed via code review???

martiniil commented on pilz_control/include/pilz_control/pilz_joint_trajectory_controller_impl.h:
```diff
@@ -178,6 +180,39 @@ triggerMovementToHoldPosition()
+{
+  switchToHoldMode();
+  JointTrajectoryController::preemptActiveGoal();
+  // TODO: Update desired state
```
I will think about it. Seems tricky.

martiniil reviewed (changes requested):
See comments. I will do a more thorough review when all parts are there.

What is the concept for setting the speed limit (turning the limit on/off)?

martiniil commented on pilz_control/include/pilz_control/pilz_joint_trajectory_controller_impl.h:
```diff
@@ -178,6 +183,48 @@ triggerMovementToHoldPosition()
+  //      active_goal->preallocated_result_->error_code = control_msgs::FollowJointTrajectoryResult::PATH_TOLERANCE_VIOLATED;
+  //      active_goal->setCanceled(active_goal->preallocated_result_);
+  // Unfortunately this does not work because sometimes the goal does not seems to get cancelled.
+  // It has to be investigated why! -> Probably a threading problem.
```
Das Problem könntest du umgehen, indem du den RealTimeGoalHandlePtr zurückgibst und an die hold-Trajektorie weitergibst.
```
template <class SegmentImpl, class HardwareInterface>
void PilzJointTrajectoryController<SegmentImpl, HardwareInterface>::
reactToFailedStateCheck(const ros::Time& curr_uptime)
{
  RealtimeGoalHandlePtr gh = triggerCancellingOfActiveGoal();

  stop_traj_builder_->buildTrajectory(JointTrajectoryController::old_time_data_.uptime.toSec(),
                                      JointTrajectoryController::old_desired_state_,
                                      stop_traj_velocity_violation_.get(), gh);
```

martiniil replied:
Natürlich ist das Problem damit nicht gelöst. Es besteht genauso in der update()-Funktion bei den tolerance-checks:
```
          rt_segment_goal->setAborted(rt_segment_goal->preallocated_result_);
          rt_active_goal_.reset();
```
Auch hier keine Garantie, dass das durchgeht, bevor `rt_segment_goal` out of scope geht.

hslusarek replied:
I suggest to close this because it is not a problem caused by us and should be handled by a different PR. I added a comment above the corresponding code lines. Is this enough for you?

martiniil replied:
Yes. You could reference https://github.com/ros-controls/ros_controllers/issues/174

hslusarek replied:
> Yes. You could reference [ros-controls/ros_controllers#174](https://github.com/ros-controls/ros_controllers/issues/174)

What do you mean? Should I reference the PR in a comment in code?

hslusarek replied:
Added. -> Done.

martiniil commented:
I would suggest to revert ba2ccf6 and instead only remove the old speed observer from the launch file, such that it does not get started with operation mode support.
Then we merge 1. this PR, 2. #296, 3. #318

martiniil commented on pilz_control/test/integrationtest_pilz_joint_trajectory_controller.py:
```diff
@@ -121,155 +120,197 @@ def observeStop(self, max_deceleration = MAX_DECELERATION, timeout = STOP_DURATI
         # Make sure robot stops (not abruptly)
-        MovementObserver().observeStop()
-        rospy.loginfo("  - Stop of robot observed!")
+        MovementObserver().observe_stop()
```
I know this existed before, but ...

The robot stops already whilst the hold request is processed. So this check has to be done in parallel to the hold request, else it is meaningless.

hslusarek replied:
OK, if the robot really already stops while the hold mode is requested, I suggest to make the motion "longer". Because everything already happens in parallel. The observer simply waits until we are sure that the motion is in progress. The motion itself does not happen in the Test-Framework but in the ROS system in parallel. If we are sure that a motion happens, we trigger the hold. A thread does not improve this situation, only a longer motion, so that we have enough time to react.

hslusarek replied:
Done.

martiniil commented on pilz_control/test/integrationtest_pilz_joint_trajectory_controller.py:
```diff
@@ -121,155 +120,197 @@ def observeStop(self, max_deceleration = MAX_DECELERATION, timeout = STOP_DURATI
+                        "Motion to default position failed")
+
+        self.assertTrue(self._move_to(position=far_away_position, duration=0.1),
+                        "Cartesian speed monitor did not detect speed limit violation")
```
This checks if the result is `SUCCESSFUL`, so you should use `assertFalse`.

hslusarek replied:
No. This is the problem: Even so the motion is cancelled, the goal still returns `SUCCESSFUL` as `error_code`, therefore, I check for `true` here. This problem is due to a strange implementation in the goal. We talked about this. I can, however, add a comment.

martiniil replied:
Ok. I did not see that.
In this case I would separate the `_move_to` call and the assertion such that the latter can be commented out and add a small comment.

hslusarek replied:
Done.

martiniil commented on pilz_control/test/integrationtest_pilz_joint_trajectory_controller.py:
```diff
@@ -121,155 +120,197 @@ def observeStop(self, max_deceleration = MAX_DECELERATION, timeout = STOP_DURATI
+
+        self.assertEqual(GoalStatus.PREEMPTED, self._trajectory_dispatcher.get_last_state(), 'Goal was not preempted.')
+
+    def test_speed_monitoring(self):
```
I think this should be extended by moving different joints. Else a robot model with a single continuous joint would suffice.

martiniil replied:
Of course, we can discuss this in common.

hslusarek replied:
> I think this should be extended by moving different joints. Else a robot model with a single continuous joint would suffice.

The model has two joints, however, the `RobotMock` class only offers one joint. I tried to come up with a generally usable dummy model and, therefore, added a  second joint. In this case, however, the second joint is not used, therefore, from the tests perspective we actually have a robot model with only one continuous joint.

I also thought, that the test that all links are monitored, is already covered by the unit test. Therefore, I see no need to repeat this test here.

martiniil replied:
Hm. The unit test covers that the `CartesianSpeedMonitor` checks all links of the model that is injected. By repeating it here, we basically would test, that the injection of the robot model is done correctly by the controller.

martiniil replied:
I like the idea of the generally usable dummy model. I think the `RobotMock` should be tailored to it. What do you think?

hslusarek replied:
> I like the idea of the generally usable dummy model. I think the `RobotMock` should be tailored to it. What do you think?

OK. I will integrate the second joint into the `RobotMock` class.

hslusarek replied:
Done.

martiniil reviewed (changes requested):
My review comments for the pjtc integrationtest:

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor_exception.h:
```diff
@@ -0,0 +1,38 @@
+#define PILZ_CONTROL_CARTESIAN_SPEED_MONITOR_EXCEPTION_H
+
+#include <stdexcept>
+
```
`#include <string>` missing.

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor_exception.h:
```diff
@@ -0,0 +1,38 @@
+/**
+ * @brief Exception thrown by the CartesianSpeedMonitor class.
+ */
+class CartesianSpeedMonitorException : public std::runtime_error
```
Bad naming, e.g. do not use `Exception` for an exception, instead describe the error case. In this case I suggest to rename it to something like `NumberOfJointsMismatch`. If you do this you can also hard code your error string. If somebody later on needs an exception base class, it still can be introduced, but right now nobody needs it.

For some hints on how to name exceptions see for example [Kevlin Henney on Youtube](https://youtu.be/ZsHMHukIlJY?t=1730).

martiniil replied:
Thanks. Done.

hslusarek commented on pilz_control/src/cartesian_speed_monitor.cpp:
```diff
@@ -0,0 +1,107 @@
+  auto p2_cart {desired_state->getGlobalLinkTransform(link)};
+
+  auto dist {(p2_cart.translation() - p1_cart.translation()).norm()};
+  auto speed {dist / time_delta};
```
All variables can be `const`.

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor_exception.h:
```diff
@@ -0,0 +1,38 @@
+{
+public:
+  CartesianSpeedMonitorException(const std::string &what_arg)
+    : std::runtime_error(what_arg)
```
I'm not sure... Maybe it is more a [invalid_argument](https://en.cppreference.com/w/cpp/error/invalid_argument) exception. What do you think?

martiniil replied:
Agreed. Done.

hslusarek commented on pilz_control/src/cartesian_speed_monitor.cpp:
```diff
@@ -0,0 +1,108 @@
+//! @return True if all parent joints in the chain below the specified link are fixed, otherwise false.
+bool hasOnlyFixedParentJoints(const moveit::core::LinkModel * const &link)
+{
+  auto parent_link {link};
```
Please re-check the variables in the `CartesianSpeedMonitor` functions, a lot of them can and should be `const`.

martiniil replied:
Done.

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor.h:
```diff
@@ -0,0 +1,92 @@
+   *
+   * @returns The computed cartesian speed.
+   */
+  static double linkSpeed(const robot_state::RobotStateConstPtr& current_state, // !!! Keep this RobotStateConstPtr for
```
Why is this function `public`? Only for the tests? If so, why not simply move this function out of the `CartesianSpeedMonitor` class, like done with the `hasOnlyFixedParentJoints()` function. If the function is moved out of the class, the testing of the `linkSpeed` function would even be simplified.

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor.h:
```diff
@@ -0,0 +1,92 @@
+   * @param link
+   * @param time_delta Time for reaching the desired state.
+   *
+   * @returns The computed cartesian speed.
```
Choose between `@brief` and `@returns` because this is simply a repetition of information already given.

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor.h:
```diff
@@ -0,0 +1,92 @@
+  //! @brief The robot states are kept in order to allow efficient getGlobalLinkTransform calls
+  robot_state::RobotStatePtr state_desired_;
+
+  std::vector<std::string> joint_names_;
```
Should be a `const` vector.

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor.h:
```diff
@@ -0,0 +1,92 @@
+   *
+   * @param current_state
+   * @param desired_state
+   * @param link
```
Remove or add useful docu.

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor.h:
```diff
@@ -0,0 +1,92 @@
+   * @param current_state
+   * @param desired_state
+   * @param link
+   * @param time_delta Time for reaching the desired state.
```
Unit of `time_delta`?

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor.h:
```diff
@@ -0,0 +1,92 @@
+   * @param current_position Current positions of controlled joints.
+   * @param desired_position Desired positions of controlled joints.
+   * @param time_delta Time for reaching the desired positions.
+   * @param speed_limit
```
In which units are these arguments stated?

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor.h:
```diff
@@ -0,0 +1,92 @@
+  /**
+   * @brief Constructor.
+   *
+   * @param joint_names List of joint names of controlled joints. The same order has to be used for joint positions.
```
Who checks this: ` List of joint names of controlled joints. The same order has to be used for joint positions.`?
This is an error, waiting to happen...

martiniil replied:
This is a convention in the `JointTrajectoryController`. A violation would result in wrong speed values.
The `CartesianSpeedMonitor` cannot verify, if the position values are ordered correctly.

hslusarek commented on pilz_control/test/unittest_cartesian_speed_monitor.cpp:
```diff
@@ -0,0 +1,255 @@
+  // the origin of link3 moves slower than the origin of link2
+  builder.addChain("link2->link3", "fixed", { origin3 });
+
+  model_ = builder.build();
```
Move to sub-function. Maybe a stand-alone static factory function?

martiniil replied:
Done.

hslusarek commented on pilz_control/test/unittest_cartesian_speed_monitor.cpp:
```diff
@@ -0,0 +1,255 @@
+
+/**
+ * @tests{Monitor_Speed_of_all_links_until_TCP,
+ * Tests speed monitoring of all moveable links.
```
Useless if always equal for each test. Summarize in one sentence what each test does, in this case maybe something like `Tests that no motion are correctly detected.`

Please update the other test summaries accordingly.

martiniil replied:
Done.

hslusarek commented on pilz_control/src/cartesian_speed_monitor.cpp:
```diff
@@ -0,0 +1,108 @@
+}
+
+double CartesianSpeedMonitor::linkSpeed(const robot_state::RobotStateConstPtr& current_state, // Important! Keep this RobotStateConstPtr to allow efficient
+                                        const robot_state::RobotStateConstPtr& desired_state, // getGlobalLinkTransform calls
```
We do not participate in the life-time management, of the `current_state` and the `desired_state` variables, so why not simply use the standard pointer `*` as argument?

And whats with the information given in the comment?The comment is useless, if it is not stated what exactly was tried.

martiniil replied:
I updated the comment.

What is the benefit of switching from shared to standard pointers here?

hslusarek replied:
Ok, decide for yourself. Have a look at the C++ core guidelines ([summary regarding smart pointers on modernescpp.com](https://www.modernescpp.com/index.php/c-core-guidelines-passing-smart-pointer)). Do one of the rules apply to us here? I don't think so. We do not want to participate in the life time management, therefore, from a Clean Code perspective I suggest using raw pointers to indicate that we do not participate in the life time management. Here is the actual C++ core guideline rule on [GitHub](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#r30-take-smart-pointers-as-parameters-only-to-explicitly-express-lifetime-semantics).

martiniil replied:
Done.

hslusarek commented on pilz_control/test/unittest_cartesian_speed_monitor.cpp:
```diff
@@ -0,0 +1,255 @@
+  state->setToDefaultValues();
+  state->updateLinkTransforms();
+  moveit::core::LinkModel* link = model_->getLinkModel("link2");
+  double time_delta = 1.0;
```
Please re-check the `const`-ness of all your variables. This one for example can be `const`.

martiniil replied:
Done.

hslusarek commented on pilz_control/test/unittest_cartesian_speed_monitor.cpp:
```diff
@@ -0,0 +1,255 @@
+  robot_state::RobotStatePtr state2(new robot_state::RobotState(model_));
+  state2->setToDefaultValues();
+  double angular_displacement{ 0.1 };
+  state2->setVariablePosition(joint_names_[1], angular_displacement);
```
Use `at(1)` for container access, to improve debugging information, in case something goes wrong.

Please use `at()` instead of `[]` wherever possible.

martiniil replied:
Done.

hslusarek commented on pilz_control/test/unittest_cartesian_speed_monitor.cpp:
```diff
@@ -0,0 +1,255 @@
+  link = model_->getLinkModel("link3");
+  speed = CartesianSpeedMonitor::linkSpeed(state, state2, link, time_delta);
+  EXPECT_NEAR(speed, expected_velocity, SPEED_COMPARISON_TOLERANCE);
+}
```
Code duplication in test `testLinkSpeedMoveJoint2()`. I suggest to simply use a loop or use [gtest's data-driven test features](https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#value-parameterized-tests).
What do you want to do, if the model consists of more than 2 joints? Do you write a new test for each joint?

martiniil replied:
See below.

hslusarek commented on pilz_control/test/unittest_cartesian_speed_monitor.cpp:
```diff
@@ -0,0 +1,255 @@
+ * Tests speed monitoring of all moveable links.
+ * }
+ */
+TEST_F(CartesianSpeedMonitorTest, testLinkSpeedLinearInTime)
```
Test intention not clear. Please, describe more clearly what you want to archive/check with this test.

martiniil replied:
Done.

hslusarek commented on pilz_control/test/unittest_cartesian_speed_monitor.cpp:
```diff
@@ -0,0 +1,255 @@
+  limit = angular_displacement / time_delta + SMALL_SKIP;
+  EXPECT_FALSE(
+      monitor.cartesianSpeedIsBelowLimit({ 0.0, 0.5 * M_PI }, { angular_displacement, 0.5 * M_PI }, time_delta, limit));
+}
```
Code duplication in test `testBelowLimitMoveJoint2()`. Like before, I suggest to simply use a loop or use [gtest's data-driven test features](https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#value-parameterized-tests).

I'm aware of the fact that there are two additional lines in the `testBelowLimitMoveJoint1()` test. Maybe use the new `robot_mock.urdf` model, to unify both tests. What do you think? I can also refactor the tests to use the new test model.

martiniil replied:
I reduced the code duplication by using a for-loop.

I would keep the model. It intentionally has a chain with two continuous joints. When moving a single joint, we expect different outcome for different joints. Especially, we can cover two cases: 2 Links are fixed in perpendicular position, or not.
This would not be possible with the new model.

hslusarek commented on pilz_control/src/cartesian_speed_monitor.cpp:
```diff
@@ -0,0 +1,108 @@
+      monitored_links_.push_back(link);
+      ROS_INFO_STREAM("Monitoring cartesian speed of link " << link->getName());
+    }
+  }
```
Move to static sub-function for better readability?

martiniil replied:
I would keep it.

hslusarek commented on pilz_control/test/unittest_cartesian_speed_monitor.cpp:
```diff
@@ -0,0 +1,255 @@
+ * Tests speed monitoring of all moveable links.
+ * }
+ */
+TEST_F(CartesianSpeedMonitorTest, testBelowLimitMoveJoint2)
```
Refactor test name and test docu, it is not very meaningful because right now I would ask: Why are there two tests (`testBelowLimitMoveJoint1`and `testBelowLimitMoveJoint2`), basically doing the same thing? => Explain in more detail your test intention.

martiniil replied:
I refactored the tests.

hslusarek commented on pilz_control/test/unittest_cartesian_speed_monitor.cpp:
```diff
@@ -0,0 +1,255 @@
+ * Tests speed monitoring of all moveable links.
+ * }
+ */
+TEST_F(CartesianSpeedMonitorTest, testLinkSpeedMoveJoint1)
```
Refactor test name and test docu, it is not very meaningful because right now I would ask: Why are there two tests (`testLinkSpeedMoveJoint1`and `testLinkSpeedMoveJoint2`), basically doing the same thing? => Explain in more detail your test intention.

martiniil replied:
I refactored the tests

hslusarek reviewed (commented):
Code review for CartesianSpeedMonitor class and CartesianSpeedMonitor unit-tests.

hslusarek commented on pilz_control/test/unittest_traj_mode_state_machine.cpp:
```diff
@@ -0,0 +1,87 @@
+  bool found_index{false};
+  for (unsigned int i = 0; i < NUM_MODES; ++i)
+  {
+    if (isTransitionValid(mode, i))
```
If find the usage of the function `isTransitionValid()` at this line very misleading. Because you are not checking if the transition is valid. You actually are checking if the current index belongs to the `hold` mode. This is a totally other thing.
Looking at the other two tests I suggest, to simply introduce a simple search function, which returns the index for a given mode. After the execution of the search function you can check if the corresponding transition is possible. This way your test also becomes more easily understandable and readable.

martiniil replied:
Actually, I find the function `isTransitionValid` misleading. A transition is defined by two states (modes). I would suggest to refactor it such that it takes the current index instead of the new index. What do you think?

The search function for the tests makes sense.

hslusarek replied:
>  I would suggest to refactor it such that it takes the current index instead of the new index.

That makes no sense to me. I want to test that my intended transition from one mode to another is valid. Therefore, I have to know the current mode is and what the requested new mode is. Using the current_index and the mode belonging to the current_index, I'm missing the information which transition is requested and, therefore, the check makes no sense...

martiniil replied:
> Therefore, I have to know the current mode and the requested new mode.

That is my point. Currently the function takes the requested new mode plus the new mode index.

hslusarek replied:
I decided to change the code completely. I simplified the state machine handling drastically. Now, only the function `isTransitionValid()` exists which takes the current mode and the requested mode and states if a transition exists or not. Hopefully, this is what you wanted.

hslusarek commented on pilz_control/test/unittest_traj_mode_state_machine.cpp:
```diff
@@ -0,0 +1,87 @@
+{
+  for (unsigned int i = 0; i < NUM_MODES; ++i)
+  {
+    auto mode = getMode(i);
```
Please re-check the `const`-ness of your variables in all functions.

martiniil replied:
Done.

hslusarek commented on pilz_control/test/unittest_traj_mode_manager.cpp:
```diff
@@ -0,0 +1,191 @@
+  return wait_future.wait_for(std::chrono::seconds(timeout)) == std::future_status::ready;
+}
+
+TEST(TrajModeManagerTest, testInitialModeStopping)
```
Looking at the naming of the other tests and the thinks you are actually doing in this test, I suggest to name this test in accordance to the other tests e.g. `testStoppingMode`

martiniil replied:
Done.

hslusarek reviewed (commented):
Review for tests: 
- `unittest_traj_mode_listener.cpp`
- `unittest_traj_mode_manger.cpp`
- `unittest_traj_mode_state_machine.cpp`

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor.h:
```diff
@@ -0,0 +1,92 @@
+   * @brief Check if cartesian speed of all links is below the speed limit.
+   *
+   * @param current_position Current positions of controlled joints.
+   * @param desired_position Desired positions of controlled joints.
```
In the constructor it is mentioned that "[...] joint names of controlled joints. The same order has to be used for joint positions". It think a note should state this fact here, too. What do you think?

Also state the unit in which the positions have to be stated.

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor.h:
```diff
@@ -0,0 +1,92 @@
+   * @param current_position Current positions of controlled joints.
+   * @param desired_position Desired positions of controlled joints.
+   * @param time_delta Time for reaching the desired positions.
+   * @param speed_limit
```
Document the `speed_limit` argument. Which links are considered etc. Does the `speed_limit` apply to all links, etc.?

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor.h:
```diff
@@ -0,0 +1,92 @@
+{
+public:
+  /**
+   * @brief Constructor.
```
Redundant. Already stated by doxygen itself...

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor.h:
```diff
@@ -0,0 +1,92 @@
+  CartesianSpeedMonitor(const std::vector<std::string> &joint_names,
+                        const robot_model::RobotModelConstPtr &kinematic_model);
+
+   //! @brief Initialize speed monitor by collecting all moveable links.
```
In my opinion the description should be kept more general and focus on the important facts, something like:
`Prepares the CartesianSpeedMonitor for execution.` The specific details what exactly happens can be read from the code and should be easy if code readability is ensured.

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor.h:
```diff
@@ -0,0 +1,92 @@
+  void init();
+
+  /**
+   * @brief Check if cartesian speed of all links is below the speed limit.
```
What is meant by `all links`? Please state exactly which links are checked and which are not checked.

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor.h:
```diff
@@ -0,0 +1,92 @@
+
+  std::vector<std::string> joint_names_;
+
+  //! @brief All moveable robot links are monitored
```
I think you mean something like: `Stores all monitored links.` This statement is strange because it makes no statement about the member variable but about the `CartesianSpeedMonitor` class itself.

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor.h:
```diff
@@ -0,0 +1,92 @@
+namespace pilz_control
+{
+
+//! @brief Monitors the cartesian speed of all moving links of a position-controlled robot.
```
Please explain in more detail `all moving links` or simply remove it because you explain it later on, too.

martiniil replied:
Actually, `monitor all moving links` is redundant/implementation detail and can be reduced to `monitor all links`.
Done.

hslusarek commented on pilz_control/include/pilz_control/cartesian_speed_monitor_exception.h:
```diff
@@ -0,0 +1,38 @@
+namespace pilz_control
+{
+/**
+ * @brief Exception thrown by the CartesianSpeedMonitor class.
```
Add more precise and useful documentation by stating exactly when this exception should be thrown.

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/pilz_joint_trajectory_controller.h:
```diff
@@ -113,20 +147,79 @@ class PilzJointTrajectoryController
 
+private:
+    /**
+     * @brief Invoke cartesian speed monitoring. Perform controlled stop in case of speed limit violation.
```
I suggest to connect both sentence both sentence via "and".

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/pilz_joint_trajectory_controller.h:
```diff
@@ -113,20 +147,79 @@ class PilzJointTrajectoryController
+     *
+     * @param period The time passed since the last update.
+     *
+     * @returns False if the cartesian speed monitoring is active and the speed limit violated, otherwise true.
```
Why so complicated? Simply write something like: "returns false if one or more links violate the Cartesian speed limit, otherwise true.

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/pilz_joint_trajectory_controller.h:
```diff
@@ -113,20 +147,79 @@ class PilzJointTrajectoryController
+    //! The currently max allowed speed for each frame on the Cartesian trajectory.
+    std::atomic<double> cartesian_speed_limit_ {0.0};
+
+    //! A robot model is needed for the cartesian speed monitor. The loader must not be destroyed before it.
```
Not really understandable...

 Firstly, what do you mean? Do I need the `RobotModelLoader` as long as the `CartesianSpeedMonitor` exists? The text does not really reflect this fact.

Secondly, why does the `RobotModelLoader` needs to exist so long? Is it not enough to store the `RobotModel` as long as the `CartesianSpeedMonitor` exists?

martiniil replied:
I added more detailed explanation.

hslusarek commented on pilz_control/include/pilz_control/traj_mode_manager.h:
```diff
@@ -0,0 +1,249 @@
+  void waitForMode();
+  //! @returns true if the given mode corresponds to the target mode, otherwise false.
+  bool isTargetModeReached(const TrajProcessingMode& mode) const;
+  void triggerListener();
```
Docu missing.

martiniil replied:
Done.

martiniil replied:
I would suggest to rename `isTargetModeReached`->`isTargetMode`, because that describes exactly method does. The name `isTargetModeReached` originates from the context where the method is called, which makes it hard to understand. What do you think?

hslusarek replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/traj_mode_manager.h:
```diff
@@ -0,0 +1,249 @@
+public:
+   //! @param mode Mode to wait for.
+  TrajProcessingModeListener(const TrajProcessingMode& mode);
+  void waitForMode();
```
Docu missing.

hslusarek commented on pilz_control/include/pilz_control/traj_mode_manager.h:
```diff
@@ -0,0 +1,249 @@
+inline static constexpr unsigned int getNextIndex(const unsigned int& current_idx)
+{
+  return (current_idx+1) % NUM_MODES;
+}
```
Add documentation? What do you think? Necessary or not?

martiniil replied:
Not necessary.

hslusarek commented on pilz_control/include/pilz_control/traj_mode_manager.h:
```diff
@@ -0,0 +1,249 @@
+  bool switchTo(const TrajProcessingMode& mode, const bool success_at_transition_only=true);
+  bool setMode(const TrajProcessingMode& mode, const bool& success_at_transition_only);
+  void callListener(const TrajProcessingMode& mode);
+  TrajProcessingMode getCurrentModeLockFree() const;
```
Docu missing?

martiniil replied:
When trying to add docu I noticed that I cannot explain the difference between `switchTo` and `setMode` in a meaningful way. I know the difference from looking at the code. It does not make much sense to me. I would suggest to refactor `setMode` into a "standard" setter function in order to improve readability.

Sorry for not carefully reviewing this earlier.

More problems I see:
- potential threading problem between the two function calls:
```
  if (setMode(mode, success_at_transition_only))
  {
    callListener(mode);
```
Do you think we should protect this execution by acquiring a lock?

- The parameter `success_at_transition_only` seems redundant to me and should be omitted.

hslusarek replied:
> potential threading problem between the two function calls:

I rather suggest that we separate the call to the listeners to an extra thread which has an input queue. This way no state switch message will be lost and no threading problem should occur, given that the input queue of the thread is thread safe. 
If you introduce an additional lock, you will probably make room for a deadlock situation because both functions operate with a different lock.

hslusarek replied:
> The parameter success_at_transition_only seems redundant to me and should be omitted.

Why is it redundant? The function `unholdEvent()` needs it, to operate correctly.

hslusarek replied:
> When trying to add docu I noticed that I cannot explain the difference between switchTo and setMode in a meaningful way. I know the difference from looking at the code. It does not make much sense to me. I would suggest to refactor setMode into a "standard" setter function in order to improve readability.

The decision to separate the `setMode()` and the `callListener()` method (which basically lead to the `switchTo()` method), has a lot to do with threading and deadlocks. You can of course suggest something else, but please be careful with the two locks because if you mix them, you will probably make room for deadlocks.

martiniil replied:
> > The parameter success_at_transition_only seems redundant to me and should be omitted.
> 
> Why is it redundant? The function `unholdEvent()` needs it, to operate correctly.

Ok, but why so complicated? The caller of `unholdEvent()` has the possibility to ask `isUnhold()` beforehand. I think you could reduce the complexity of the code by making these event functions more consistent and get rid of `success_at_transition_only`. Decide for yourself.

martiniil replied:
> > potential threading problem between the two function calls:
> 
> I rather suggest that we separate the call to the listeners to an extra thread which has an input queue. This way no state switch message will be lost and no threading problem should occur, given that the input queue of the thread is thread safe.
> If you introduce an additional lock, you will probably make room for a deadlock situation because both functions operate with a different lock.

Hm. I am not really glad with the fact, that whilst one event is processed, another one can be processed in between. The input queue also does not prevent this. As far as I understand the point of a state machine is to prevent such a thing.
Maybe we should rethink the state machine used here. I suggest to consider:
![IMG_9193](https://user-images.githubusercontent.com/35950815/77749718-c6849c80-7022-11ea-8de5-377de099fcca.JPG)

hslusarek replied:
> > > The parameter success_at_transition_only seems redundant to me and should be omitted.
> > 
> > 
> > Why is it redundant? The function `unholdEvent()` needs it, to operate correctly.
> 
> Ok, but why so complicated? The caller of `unholdEvent()` has the possibility to ask `isUnhold()` beforehand. I think you could reduce the complexity of the code by making these event functions more consistent and get rid of `success_at_transition_only`. Decide for yourself.

Because the `unholdEvent()` function has to be an atomic operation which includes the `isUnhold()` function call. The `isUnhold()` call cannot be performed externally because then it the complete operation is not atomic.

hslusarek commented on pilz_control/include/pilz_control/traj_mode_manager.h:
```diff
@@ -0,0 +1,249 @@
+  bool isUnhold();
+  void registerListener(TrajProcessingModeListener* const listener);
+
+  TrajProcessingMode getCurrentMode();
```
Docu missing.

hslusarek commented on pilz_control/spec.dox:
```diff
@@ -0,0 +1,56 @@
+ * @subsubsection speed_monitoring Speed Monitoring
+ *
+ * @definereq{Monitor_Speed_of_all_links_until_TCP}
+ *  - The system monitors all links of the robot (including the TCP) in order to determine their
```
What is meant by "all links"? Please describe exactly which links are monitored and which aren't. What happens for example with gripper links?

martiniil replied:
Done.

hslusarek commented on pilz_control/spec.dox:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @subsubsection controller_hold Controller Hold
+ *
+ * @definereq{no_execution_during_hold}
```
The specifications could be linked against unit-tests. Yes, right now they cannot be linked against the integration test (Python) but this should not stop us from linking against the unit-tests (C++). What do you think?

martiniil replied:
I would omit the unit-tests from the docu. That is why I had to remove the referencing, too.

hslusarek replied:
> I would omit the unit-tests from the docu. That is why I had to remove the referencing, too.

I only reluctantly accept this statement because our "inability" to do it right, should not lead to the effect of destroying the main purpose of the specification: Giving inside why a test exists and showing if a test is still relevant or not, by showing the corresponding requirement.

martiniil replied:
You could see it like follows: The requirement actually is captured in the test documentation. What is missing at the moment is an automated build routine, that generates the links and uploads the specification to a place where the whole team can read it.
I would say that this should not lead to the effect of polluting the code documentation on RosWiki.

hslusarek commented on pilz_control/include/pilz_control/pilz_joint_trajectory_controller_impl.h:
```diff
@@ -17,13 +17,34 @@
 namespace ph = std::placeholders;
 
+template<class Segment>
+bool isTrajectoryExecuted(const std::vector< TrajectoryPerJoint<Segment>>& traj,
```
Docu missing?

martiniil replied:
Done.

hslusarek commented on prbt_hardware_support/include/prbt_hardware_support/operation_mode_setup_executor.h:
```diff
@@ -63,16 +56,11 @@ class OperationModeSetupExecutor
   //! The active speed override
   std::atomic<double> speed_override_{0.0};
 
   //! Function used to propagate speed limit changes into the system.
```
Update docu?

martiniil replied:
Done.

hslusarek reviewed (commented):
Code docu review.

I have to admit that I'm not totally sure how much documentation is actually needed. If a function is well named and easily readable, I'm tend to suggest that no addtional documentation is needed. But this is just my opinion. I added a comment whenever a documentation was missing. However, I leave it to you, to decide if you actually want to add additional documentation or not.

hslusarek commented on prbt_hardware_support/src/operation_mode_setup_executor.cpp:
```diff
@@ -40,25 +38,25 @@ void OperationModeSetupExecutor::updateOperationMode(const OperationModes& opera
   }
 
-  if (set_speed_limit_func_)
+  if (monitor_cartesian_speed.is_initialized() && monitor_cartesian_speed_func_)
```
I think instead of `monitor_cartesian_speed.is_initialized()`, you can simply write `monitor_cartesian_speed`.

martiniil replied:
Done.

hslusarek commented on prbt_hardware_support/test/unittests/unittest_operation_mode_setup_executor.cpp:
```diff
@@ -209,48 +208,48 @@ TEST_F(OperationModeSetupExecutorTest, testUpdateOperationModeUnknownMode)
 };
 
-MATCHER_P(IsCorrectSpeedLimitSet, speed_limit, "") { return arg.request.speed_limit == speed_limit; }
+MATCHER_P(IsMonitorCartesianSpeedCorrect, active_flag, "") { return arg.request.data == active_flag; }
```
Function name is not really fitting. I suggest something like: `isSpeedMonitoringSettingCorrect`.

martiniil replied:
Done.

hslusarek commented on prbt_hardware_support/test/unittests/unittest_disable_speed_monitoring.cpp:
```diff
@@ -0,0 +1,88 @@
+
+/**
+ * @tests{speed_monitoring_without_operation_mode_support,
+ *  Tests the deactivation of the speed monitoring.
```
Please write a little bit more docu. It is extremely difficult to understand what "object/code" exactly this test tries to test. Write something like: "Tests the deactivation of the speed monitoring via service call from `fake_operation_mode_setup.launch` file."

martiniil replied:
Done.

hslusarek reviewed (commented):
Code review for:
- `unittest_disable_speed_monitoring`.

hslusarek commented on pilz_control/include/pilz_control/pilz_joint_trajectory_controller_impl.h:
```diff
@@ -226,8 +213,10 @@ updateFuncExtensionPoint(const typename JointTrajectoryController::Trajectory& c
     {
-      mode_->stoppingEvent();
-      stopMotion(time_data.uptime);
+      if (mode_->stopEvent())
```
Condition can be integrated into `if` above via `&&`.

martiniil replied:
Done.

hslusarek commented on pilz_control/include/pilz_control/traj_mode_manager.h:
```diff
@@ -62,65 +62,57 @@ class TrajProcessingModeStateMachine
-  //! @returns true if the given mode corresponds to the target mode, otherwise false.
-  bool isTargetMode(const TrajProcessingMode& mode) const;
-  //! @brief Notify the listener that the target mode is reached.
+  void waitForHold();
```
I thought about it a little bit. I think the previous and the new name are actually not good. Let's revisite what Kevlin Henney says in his talk, see on [Youtube](https://youtu.be/ZsHMHukIlJY?t=1614). I suggest to simply use `wait`. For what else would we use a `HoldModeListener`?

martiniil replied:
Sounds good. Done.

hslusarek commented on pilz_control/include/pilz_control/pilz_joint_trajectory_controller_impl.h:
```diff
@@ -146,28 +145,16 @@ template <class SegmentImpl, class HardwareInterface>
 bool PilzJointTrajectoryController<SegmentImpl, HardwareInterface>::
 handleUnHoldRequest(std_srvs::TriggerRequest&, std_srvs::TriggerResponse& response)
```
I find the different behavior of the two service-functions `handleHoldRequest()` and `handleUnHoldRequest()` very confusing. `handleHoldRequest()` does wait until the stop is finished. In other words, the service is blocking. `handleUnHoldRequest()`, however, does not wait until the `unhold` mode is reached but returns immediately. In other words, the service is a non-blocking service. From a user perspective this is very confusing and not intuitive.

hslusarek replied:
For me the fact that `handleUnHoldRequest()` does not longer wait is a major change. Therefore, this has to be tested at the real system.

martiniil replied:
I can see that it is confusing from this perspective. However this is what I would expect from a Hold/Unhold function: (forget our hold-/unhold requests for a moment)
- "Hold":
  - can be triggered at any time,
  - if the robot is moving, the hold process triggers a stop motion and takes up time until the robot is stopped
- "Unhold":
  - can not be triggered at any time,
  - if it can be triggered, it has an immediate effect (no need to wait for the robot).

What do you think about that?

In other words:
"If I cannot execute "Unhold" currently, the service takes longer ..."
or
"If I cannot execute "Unhold" currently, try next time ..."

hslusarek replied:
To be honest, I think the connection type is wrong. At least the `hold` function (probably the `unhold` function, too) should be provided via [action](http://wiki.ros.org/actionlib). If an action is used as connection type, many of the problems discussed would not exist. Furthermore, the user interface would be easier to understand and would leave the decision to wait or not, to the user.

martiniil replied:
Let's continue the discussion: https://github.com/PilzDE/pilz_robots/issues/328

hslusarek commented on pilz_control/include/pilz_control/traj_mode_manager.h:
```diff
@@ -62,65 +62,57 @@ class TrajProcessingModeStateMachine
-  //! @brief Protects the access to the listener list.
-  std::mutex listener_mutex_;
+  std::list<HoldModeListener*> listener_;
+  //! @brief Protects the access to member variables.
```
To general because e.g. it does not protect access to `mode_state_machine_`. I suggest specifically name the member which are protected. Or even better would be a mutex name which already indicates what it protects.

martiniil replied:
Actually, it protects `mode_state_machine_` indirectly, because it is used to make all public methods mutually exclusive. I changed the description to make that clear.

hslusarek commented on pilz_control/include/pilz_control/traj_mode_manager.h:
```diff
@@ -62,65 +62,57 @@ class TrajProcessingModeStateMachine
+   * @return True if transition was performed, otherwise false.
+   */
+  bool switchTo(const TrajProcessingMode& mode);
+  void registerHoldListener(HoldModeListener* const listener);
```
The name is redundant because the argument type already indicates that only `HoldModeListener` can be registered.

martiniil replied:
Done.

hslusarek reviewed (commented):
Code review for trajectory mode state machine.

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -23,13 +27,16 @@
 #include <pilz_control/pilz_joint_trajectory_controller.h>
 
-typedef hardware_interface::JointCommandInterface HWInterface;
+typedef hardware_interface::PositionJointInterface HWInterface;
 typedef trajectory_interface::QuinticSplineSegment<double> Segment;
 typedef pilz_joint_trajectory_controller::PilzJointTrajectoryController<Segment, HWInterface> Controller;
 typedef std::shared_ptr<Controller> ControllerPtr;
```
Please use `using` instead of `typedef`. For more information see for example [nextptr.com](https://www.nextptr.com/tutorial/ta1193988140/how-cplusplus-using-or-aliasdeclaration-is-better-than-typedef).

martiniil replied:
Done.

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -38,27 +45,107 @@ typedef control_msgs::FollowJointTrajectoryGoal GoalType;
+  return goal.trajectory.points.back().time_from_start;
+}
+
+static bool waitForUsingSystemTime(const std::function<bool()>& is_condition_fulfilled,
```
This function name is totally confusing. Why do I care about the system time? I just want to wait, right? The rest is implementation detail and can maybe stated in the docu but not in the name. (My opinion.) I would suggest something like: `waitFor(...)`.

martiniil replied:
Done.

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -112,191 +211,395 @@ void PilzJointTrajectoryControllerTest::SetUp()
 
   ASSERT_TRUE(ros::service::exists(controller_nh_.getNamespace() + HOLD_SERVICE, true));
   ASSERT_TRUE(ros::service::exists(controller_nh_.getNamespace() + UNHOLD_SERVICE, true));
   EXPECT_TRUE(ros::service::exists(controller_nh_.getNamespace() + IS_EXECUTING_SERVICE, true));
+
+  EXPECT_TRUE(waitForActionServer());
```
Why does `is_executing_service` and the action server connection use `EXPECT_TRUE`, while `hold` and `unhold` service use `ASSERT_TRUE`?  
I think there is a misunderstanding when it comes to using `ASSERT_` and `EXPECT_`. Please have a look at [stackoverflow](https://stackoverflow.com/a/2565309) or [Google's test blog](https://testing.googleblog.com/2008/07/tott-expect-vs-assert.html). 
- `ASSERT_` indicates that it makes no sense to continue the test, in case the check fails. 
- `EXPECT_` indicates that the test can be continued because valuable information can still be obtained by the _test_.

Here, however, the differentiation seems to be made because of the importance of the services, etc. This, however, is not the intended differentiation between `ASSERT_` and `EXPECT_`.

martiniil replied:
Oh, that's a mistake. Thanks.

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -112,191 +211,395 @@ void PilzJointTrajectoryControllerTest::SetUp()
   return goal;
 }
 
+testing::AssertionResult PilzJointTrajectoryControllerTest::checkForHold()
```
I suggest something like: `isControllerInHoldMode()`. What do you think?

martiniil replied:
Sounds good.

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -112,191 +211,395 @@ void PilzJointTrajectoryControllerTest::SetUp()
+  return testing::AssertionSuccess();
+}
+
+testing::AssertionResult PilzJointTrajectoryControllerTest::checkForUnhold()
```
I suggest something like: `isControllerInUnholdMode()`. What do you think?

martiniil replied:
Sounds good.

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -112,191 +211,395 @@ void PilzJointTrajectoryControllerTest::SetUp()
-  GoalType goal {generateSimpleGoal(goal_duration)};
+/**
+ * @tests{end_holding,
+ * Tests unholding the controller.
```
You actually test the unholding to different time points. Unfortunately, this is a little bit obscured by all the code "clutter". Maybe you can mention it in the docu.

martiniil replied:
Done.

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -112,191 +211,395 @@ void PilzJointTrajectoryControllerTest::SetUp()
 }
 
+
+bool PilzJointTrajectoryControllerTest::updateControllerUntil(const std::function<bool()>& is_condition_fulfilled,
+                                                              const std::chrono::milliseconds& timeout)
+{
+  const std::chrono::system_clock::time_point start {std::chrono::system_clock::now()};
+  do
+  {
+    if (is_condition_fulfilled())
+    {
+      return true;
+    }
+    if (std::chrono::system_clock::now() - start > timeout)
+    {
+      return false;
+    }
+    progressInTime(ros::Duration(DEFAULT_UPDATE_PERIOD_SEC));
+    updateController();
+    std::this_thread::sleep_for(std::chrono::milliseconds(SLEEP_TIME_MSEC));
+  }
+  while (ros::ok());
+  return false;
+}
```
This function is basically a duplication of the function `waitForUsingSystemTime()` except for the two function calls `progressInTime()` and `updateController()`. Is it not possible to generalize one of the two functions and avoid this code duplication?

martiniil replied:
Done.

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -112,191 +211,395 @@ void PilzJointTrajectoryControllerTest::SetUp()
+                                             });
+
+  std::chrono::milliseconds hold_timeout {WAIT_FOR_HOLD_FUTURE_MSEC};
+  EXPECT_TRUE(updateControllerUntil([&hold_future](){ return isFutureReady(hold_future); }, hold_timeout));
```
The function name `updateControllerUntil()` does not really indicate that something is returned and if so what.

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -112,191 +211,395 @@ void PilzJointTrajectoryControllerTest::SetUp()
+  std_srvs::TriggerRequest req;
+  std_srvs::TriggerResponse resp;
+  // run async such that stop trajectory can be executed in the meantime
+  std::future<bool> hold_future = std::async(std::launch::async,
+                                             [this, &req, &resp]()
+                                             {
+                                               return controller_->handleHoldRequest(req, resp);
+                                             });
```
Maybe introduce a sub-function returning the future object and name it something like: `triggerAsyncStop()`. Increases readability and can be re-used in the other tests in this test suit.

martiniil replied:
Done.

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -112,191 +211,395 @@ void PilzJointTrajectoryControllerTest::SetUp()
+TEST_F(PilzJointTrajectoryControllerTest, testDoubleHoldSuccess)
 {
   ASSERT_TRUE(controller_->init(hardware_, nh_, controller_nh_)) << "Failed to initialize the controller.";
+  ASSERT_TRUE(waitForActionServer());
   controller_->state_ = controller_->INITIALIZED;
-  startController();
 
+  startController();
   controller_->state_ = controller_->RUNNING;
```
I'm confused... Why not `ASSERT_TRUE(performFullControllerStartup());` here like in `testHoldSuccessAfterUnhold`? 
Why should the first `hold` request need to wait, while the second `hold` does not. Both `hold` requests are made, while the controller is in `hold` mode after startup.

martiniil replied:
I thought, `unhold` is not necessary here and `performFullControllerStartup()` includes a `unhold`. Do you think we should still call it in order to avoid confusion?

Remember that the controller is in `stopping` mode after startup. When the first `hold` returns the controller has switched to `hold` mode.

hslusarek replied:
> Remember that the controller is in `stopping` mode after startup. When the first `hold` returns the controller has switched to `hold` mode. 

Ah, I forgot that we changed the initial mode to `stopping`. 

Anyway, what's the objective of the test? You wrote as test docu: "Tests triggering hold two times successively." If I read your docu I think that both `hold`'s are triggered while the "controller" is in idle state and not while the controller is still starting up. Therefore, I suggest to add the `performFullControllerStartup()` at the beginning.

martiniil replied:
Ok.

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -112,191 +211,395 @@ void PilzJointTrajectoryControllerTest::SetUp()
- *    2. Response contains success==True. The elapsed time corresponds to the expected stop trajectory duration.
- *    3. Response contains success==True. The second call does not trigger a stop trajectory.
+ * @tests{start_holding,
+ * Tests holding the controller.
```
The docu is a little bit too generic. Maybe be a little bit more specific because if you are not, I would ask: "What's the difference to `testHoldSuccessAfterUnhold`" which has the same docu?

martiniil replied:
You are right. thanks

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -112,191 +211,395 @@ void PilzJointTrajectoryControllerTest::SetUp()
- *
- * Increases line coverage.
+ * @tests{end_holding,
+ * Tests unholding the controller.
```
Same docu as in `testUnholdSuccess`. How can that be? Ready for deletion??? ;)

martiniil replied:
;)

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -112,191 +211,395 @@ void PilzJointTrajectoryControllerTest::SetUp()
-  updateController();
+/**
+ * @tests{end_holding,
+ * Tests unholding the controller.
```
Same docu as in `testUnholdSuccess`. How can that be? Ready for deletion??? ;)

martiniil replied:
Fixed it.

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -38,27 +45,107 @@ typedef control_msgs::FollowJointTrajectoryGoal GoalType;
+  return this_future.wait_for(std::chrono::seconds(0)) == std::future_status::ready;
+}
+
+/**
+ * @brief For simplicity this robot only updates one joint of the two joints defined in the urdf.
+ */
```
I don't really understand this explanation. Why is the already existing `robot_mock.cpp` not used. In `generateSimpleGoal()` you only command the first joint anyway, so why not use the implementation in `robot_mock.cpp`? I suggest to delete this implementation here and use `robot_mock.cpp`. If something is missing in `robot_mock.cpp`, please add it.

martiniil replied:
In this test, the controller is not launched via a `controller_spawner`, but instantiated directly in the test node. The class `RobotMock` in the unittest mocks both the `robot` and the `controller_manager`. I see no possibility of re-using `robot_mock.cpp` in this setting.

We could split up the class `RobotMock` and the `main`-function in `robot_mock.cpp`, such that the class can be used by other processes, too.
Then we replace this implementation by a "controller_manager-mock" and reuse the robot class. What do you think about that?

hslusarek replied:
> We could split up the class `RobotMock` and the `main`-function in `robot_mock.cpp`, such that the class can be used by other processes, too.
Then we replace this implementation by a "controller_manager-mock" and reuse the robot class. What do you think about that?

Sounds good.

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -434,127 +686,104 @@ TEST_P(PilzJointTrajectoryControllerIsExecutingTest, testHoldAtStart)
-  /**********
-   * Step 2 *
-   **********/
   ros::Duration default_update_period {DEFAULT_UPDATE_PERIOD_SEC};
```
I think this line is obsolete and actually clutters the test code. Simply integrate it directly in the line where the variable is used.

martiniil replied:
Done.

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -434,127 +686,104 @@ TEST_P(PilzJointTrajectoryControllerIsExecutingTest, testHoldAtStart)
-  /**********
-   * Step 2 *
-   **********/
   ros::Duration default_update_period {DEFAULT_UPDATE_PERIOD_SEC};
```
I think this line is obsolete and actually clutters the test code. Simply integrate it directly in the line where the variable is used.

martiniil replied:
Done.

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -38,27 +45,107 @@ typedef control_msgs::FollowJointTrajectoryGoal GoalType;
+}
+
 /**
  * @brief Test fixture class for the unit-test of the PilzJointTrajectoryController.
```
The `PilzJointTrajectoryControllerTest` class consists of quite a number of member variables and helper functions, partially using the member variables. Also, the `PilzJointTrajectoryControllerIsExecutingTest` is derived from `PilzJointTrajectoryControllerTest` and uses the members and the helper functions too.
Furthermore, the test suite consists of a "wall of code" which makes it difficult to understand and read.
Therefore, I suggest the following:
- Separate the members and helper functions used by both test classes (`PilzJointTrajectoryControllerTest` and `PilzJointTrajectoryControllerIsExecutingTest`) into a separate independent class.
- Put the new class into a separate header. 
- Then include the new helper-class as member in both test classes.
- Then only derive `PilzJointTrajectoryControllerIsExecutingTest` from `testing::WithParamInterface<InvokeIsExecuting>` and nothing else.

**Please note:**
- Inheritance is not the only way to re-use helper functions across different classes, etc.

martiniil replied:
I would not swap out all functions & members into a single helper class, but make it dependent on the context.
I just started implementing a `PJTCManagerMock` (see comment above). After that I will try to extract further components in order to reduced the amount of code in the test suite.

hslusarek replied:
> I would not swap out all functions & members into a single helper class, but make it dependent on the context.

No, only separate the functions & member into a helper class which actually create a logical unit. I also want to emphasize that I think that the inheritance model used here is faulty from a clean code perspective.

> After that I will try to extract further components in order to reduced the amount of code in the test suite.

OK, I will wait ;)

martiniil replied:
Done. You can recheck it if you want.

hslusarek reviewed (commented):
Code review for `unittest_pilz_joint_trajectory_controller.cpp`.

hslusarek commented on pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp:
```diff
@@ -71,232 +104,412 @@ class PilzJointTrajectoryControllerTest : public testing::Test
-  controller_->state_ = controller_->RUNNING;
+/**
+ * @tests{start_holding,
+ * Tests holding the controller.
```
You changed the docu below but not this one. Please update the docu, too.

martiniil replied:
Thanks!

hslusarek reviewed (commented):
Ignore this comment.

hslusarek commented on pilz_control/test/pjtc_test_helper.h:
```diff
@@ -0,0 +1,177 @@
+ * - Make progress in simulated ros::Time,
+ * - Invoke update_func.
+ */
+bool waitFor(const std::function<bool()>& is_condition_fulfilled,
```
Is there a reason why all functions in this header are `static` except this one?

martiniil replied:
No, I add will add it.

hslusarek reviewed (commented):
Code review: Refactoring robot_mock.cpp

Commit:
Add Cartesian speed monitoring

martiniil commented:
Controller changes are in testing right now: https://github.com/ros/rosdistro/pull/24816

martiniil reviewed (changes requested):
We still need acceptance tests. See #398.

agutenkunst commented on pilz_control/test/acceptance_test_speed_monitoring.md:
````diff
@@ -0,0 +1,66 @@
+  roslaunch pilz_control joint_states_speed_observer.launch
+  ```
+  2. Execute `rosrun pilz_control acceptance_test_speed_monitoring.py`.
+  3. Perform some movements via RViz. Choose different velocity scales.
````
Choose velocity scales (0,1] ?

agutenkunst replied:
Actually this is a bit to fuzzy for my taste. Are there edge cases that should be checked?

martiniil replied:
Yes, thanks.

agutenkunst commented on pilz_control/test/acceptance_test_speed_monitoring.md:
```diff
@@ -0,0 +1,66 @@
+### Expected Results
+  1. The robot starts properly and is moveable via Rviz.
+  2. Check the output on the console. The maxmimum frame speed should **not** have exceeded the limit of `0.25m/s`.
+  3. Same as 2.
```
Reading all this is a lot of line jumping. Feels like goto. I don't think DRY is necessary here.

agutenkunst commented on pilz_control/test/acceptance_test_speed_monitoring.md:
```diff
@@ -0,0 +1,66 @@
+
+### Test Sequence
+
+  1. Press the acknowledge button. Make sure the green light on the robot is on and mode AUTO is enabled. Run
```
We always start in T1 oder in Auto? If there needs to be a switch please explicitly state this.

martiniil replied:
@SansoneG Can you check this? Is the operation mode displayed on the power cabinet always the same at startup?

agutenkunst commented on pilz_control/test/acceptance_test_speed_monitoring.md:
```diff
@@ -0,0 +1,66 @@
+
+### Expected Results
+  1. The robot starts properly and is moveable via Rviz.
+  2. Check the output on the console. The maximum frame speed should have exceeded the limit of `0.25m/s`. The robot movements should **not** have been interrupted.
```
As far as I understand this could be done by acceptance_test_speed_monitoring.py and the tester would only have to check if the test failed.

martiniil replied:
But then the outcome depends on the operation mode, which is (currently) unknown in the test. Do you still think, that it should be done in `acceptance_test_speed_monitoring.py`?

agutenkunst commented on pilz_control/test/acceptance_test_speed_monitoring.md:
```diff
@@ -0,0 +1,66 @@
+
+### Expected Results
+  1. The robot starts properly and is moveable via Rviz.
+  2. Check the output on the console. The maxmimum frame speed should **not** have exceeded the limit of `0.25m/s`.
```
As far as I understand this could be done by acceptance_test_speed_monitoring.py and the tester would only have to check if the test failed.

agutenkunst commented on pilz_control/test/acceptance_test_speed_monitoring.md:
```diff
@@ -0,0 +1,66 @@
+  1. The robot starts properly and is moveable via Rviz.
+  2. Check the output on the console. The maxmimum frame speed should **not** have exceeded the limit of `0.25m/s`.
+  3. Same as 2.
+  4. Check the output on the console. The maximum frame speed should have exceeded the limit of `0.25m/s`. The robot movements should **not** have been interrupted.
```
As far as I understand this could be done by acceptance_test_speed_monitoring.py and the tester would only have to check if the test failed.

agutenkunst reviewed (changes requested):
Review according to https://github.com/PilzDE/pilz_robots/issues/400
some remarks.

martiniil commented:
> I know this is "only" a tool used for tests (at least for now). But why is it not tested?

So far we never wrote tests for code that is only used in tests. I think we already discussed this in another context and agreed on it.

agutenkunst commented:
> > I know this is "only" a tool used for tests (at least for now). But why is it not tested?
> 
> So far we never wrote tests for code that is only used in tests. I think we already discussed this in another context and agreed on it.

Don't agree in this case where a error of the tool would cause a wrong test result.

martiniil commented:
> Did you check this tool with a real robot where there could be more noise in the data?

Yes!

martiniil commented:
See https://github.com/ros-controls/ros_controllers/issues/493

hslusarek commented:
> See [ros-controls/ros_controllers#493](https://github.com/ros-controls/ros_controllers/issues/493)

I opened PR #405 which deactivates the command interface.

agutenkunst commented:
@martiniil I get
```
/usr/bin/ld: warning: libmoveit_robot_model.so.1.0.2, needed by /opt/ros/melodic/lib/libmoveit_robot_model_loader.so, not found (try using -rpath or -rpath-link)
/usr/bin/ld: warning: libmoveit_profiler.so.1.0.2, needed by /opt/ros/melodic/lib/libmoveit_robot_model_loader.so, not found (try using -rpath or -rpath-link)
/usr/bin/ld: warning: libmoveit_kinematics_base.so.1.0.2, needed by /opt/ros/melodic/lib/libmoveit_kinematics_plugin_loader.so, not found (try using -rpath or -rpath-link)
/opt/ros/melodic/lib/libmoveit_kinematics_plugin_loader.so: undefined reference to `kinematics::KinematicsBase::DEFAULT_SEARCH_DISCRETIZATION'
/opt/ros/melodic/lib/libmoveit_kinematics_plugin_loader.so: undefined reference to `typeinfo for kinematics::KinematicsBase'
collect2: error: ld returned 1 exit status
pilz_robots/pilz_control/CMakeFiles/joint_states_speed_observer.dir/build.make:183: recipe for target '/home/alex/catkin_ws_PBI_T1/devel/lib/pilz_control/joint_states_speed_observer' failed
make[2]: *** [/home/alex/catkin_ws_PBI_T1/devel/lib/pilz_control/joint_states_speed_observer] Error 1
CMakeFiles/Makefile2:12203: recipe for target 'pilz_robots/pilz_control/CMakeFiles/joint_states_speed_observer.dir/all' failed
make[1]: *** [pilz_robots/pilz_control/CMakeFiles/joint_states_speed_observer.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
Invoking "make -j12 -l12" failed
```

I did
```
sudo apt-get update
sudo apt-get upgrade
rosdep install --from-paths src --ignore-src --rosdistro=melodic -y
```

what did I miss?

**Update:**
- `ccache --clear`
- `rm -rf build devel`

helped

agutenkunst commented on prbt_hardware_support/test/unit_tests/unittest_brake_test_utils.cpp:
```diff
@@ -155,10 +158,16 @@ TEST(BrakeTestUtilsTest, testGetCurrentJointStates)
 
+  // Move away from zero
+  joint_states_pub.setVelocity(0.1);
+  ros::Duration(0.1).sleep();
```
sleep during tests makes them flaky. Is there no other way?

martiniil replied:
This is intended, because it will lead to a "random" joint state opposed to just reading zero values.

agutenkunst replied:
Not really happy about having random behavior inside a unittest. 
If there is no other way at least state as a comment that you want random behavior.

agutenkunst commented on pilz_testutils/include/pilz_testutils/joint_state_publisher_mock.h:
```diff
@@ -0,0 +1,103 @@
+public:
+  /**
+   * @brief Constructor.
+   * @param ns Namespace in which the joint states are to be published.
```
No param. Rather remove this comment all together if it does not state something which is not already obvious.

agutenkunst commented on pilz_testutils/include/pilz_testutils/joint_state_publisher_mock.h:
```diff
@@ -0,0 +1,103 @@
+  JointStatePublisherMock();
+
+  /**
+   * @brief Start periodic publishing in a separate thread.
```
If this is way the function does I would prefer calling it something which states this.

agutenkunst commented on pilz_testutils/include/pilz_testutils/joint_state_publisher_mock.h:
```diff
@@ -0,0 +1,103 @@
+  /**
+   * @brief Set velocity for joint1.
+   */
+  void setVelocity(const double& joint1_velocity);
```
```suggestion
  void setJoint1Velocity(const double& vel);
```

agutenkunst commented on pilz_testutils/include/pilz_testutils/joint_state_publisher_mock.h:
```diff
@@ -0,0 +1,103 @@
+  /**
+   * @brief Stop publishing joint states.
+   */
+  void stop();
```
```suggestion
  void stopPublishing();
```

agutenkunst commented on pilz_testutils/include/pilz_testutils/joint_state_publisher_mock.h:
```diff
@@ -0,0 +1,103 @@
+  /**
+   * @brief In degenerate-time-mode each message is published twice.
+   */
+  void setDegenerateTimeMode();
```
Could you provide some more explanation/reasoning of the mode?

agutenkunst replied:
Still to vague for me sorry

agutenkunst commented on pilz_control/CHANGELOG.rst:
```diff
@@ -10,7 +15,6 @@ Changelog for package pilz_control
 0.5.15 (2020-05-03)
 -------------------
 * Introduce goal_time_tolerance to PJTC function is_executing
-* Contributors: Pilz GmbH and Co. KG
```
Please leave this here....

agutenkunst commented on pilz_testutils/include/pilz_testutils/joint_state_publisher_mock.h:
```diff
@@ -0,0 +1,92 @@
+/*
```
Why does this show up as a new file? As far as I understand this is just moved from prbt_hardware_support. Did you do `git mv`?

agutenkunst commented on pilz_control/test_utils/src/joint_states_speed_observer.cpp:
```diff
@@ -0,0 +1,145 @@
+static const std::string MAX_FRAME_SPEED_TOPIC_NAME{ "max_frame_speed" };
+static const std::string MONITORED_LINK_NAMES_PARAMETER{ "/prbt/monitored_link_names" };
+
+static constexpr double T1_SPEED_LIMIT{ 0.25 };
```
Maybe remove this from this class to reduce the binding to the Norm domain?

martiniil committed:
Merge branch 'melodic-devel' into feature/speed_limit_pjtc_extend_controller

martiniil committed:
Feature/speed limit acceptance test (#407)

* Add acceptance test for speed monitoring

Use speed observer based on joint_states for testing purpose:
- Does not suffer from noise in /tf data
- Publishes maximum frame speed on /max_frame_speed

* Move joint_states_speed_observer to pilz_control/test_utils
* Add pilz_control/test_utils to coverage report
* Add roslint-test to pilz_control

* Add JointStatePublisherMock in pilz_testutils
